proposals/mk-ii.md

```diff
@@ -0,0 +1,99 @@
+# Ansible Container Mk. II
+## Architectural document
+
+### Refactor plan
+
+* The "builder container" will now be referred to as the "conductor".
+* Outside of the container on the host OS should be four and precisely four bits
+  of functionality: Building the conductor container, launching the conductor
+  container, `init`, and Project Starbuck (see below).
+* The global `setup.py` should specify an `entry_points['console_script']` that
+  invokes a super thin wrapper. All the wrapper should do is:
+    1. Parse arguments
+    2. Build the conductor container.
+      * Does this mean we need Docker for every install?
+    3. Launch the conductor container using the specified command.
+* The vast majority of all of the Ansible Container logic goes into the conductor. 
+  Like, everything. The global `setup.py` file installs this code as
+  data, not as code. There's no way to `import` the Ansible Container logic from
+  the machine `ansible-container` is installed on.
+* The build process for the conductor uses the `$PWD` as the build context.
+  To build the conductor:
+    1. Put the Ansible Container logic payload into the container.
+    2. Make a virtualenv.
+    3. Run the `setup.py` from the payload.
+    4. If a `requirements.txt` is in the build context, collect requirements.
+    5. If a `requirements.yml` is in the build context, collect requirements.
+
+### Changes to `container.yml`
+
+* All Ansible Container artifacts (`container.yml`, `requirements.yml`, etc.)
+  will be moved out of the `ansible/` directory and into the project root.
+* The `main.yml` file is no longer a thing. We got rid of it. It's gone.
+* The "fidelity to `docker-compose.yml`" preference is hereby abandoned.
+* New keys to `container.yml` services:
+  * `from` - The base image to use for builds.
+  * `roles` - An array of role references to run on top of the base image. Each
+    role will be committed as a layer on the final built image.
+* New top-level keys:
+  * `settings` which optionally contain:
+    * `engine` - the name of the preferred container engine
+    * `conductor_base` - the name of a base image from which to build your
+      conductor. The idea here is that if you're building images against a Debian
+      base, needing `python-apt` for instance, you want a Debian conductor.
+    * `vault_file` - the relative path to a vault file that should be used by
+      runs of Ansible Container in the conductor.
+
+### Changes to `engine.py`
+
+* The generic `Engine` class will cover `build`, `run`, and `shipit`.
+* The `build` process:
+  * The `virtualenv` from the conductor will be exported into each built service.
+    The Python runtime used by Ansible is this `virtualenv` - no need to have a
+    Python binary in the target containers.
```

Comment:
To be clear, this will just be *mounted* during the build step. Not included in the shipped container.

I assume this will have to be a read-only mount, right?